### ISSUE TITLE
Parse source code to add function parameters as well

### ISSUE THREAD

**Issue by tdx** [ostinelli/SublimErl#18]
How about to parse source code of project (not erlang libraries) instead of beams ?
As a bonus we can supply function parameters for completion.

**Comment by ostinelli** [comment]
what would be the advantage of this?

**Comment by tdx** [comment]
At least it's does not need to recompile code (in several situations it may be complicated task, my projects uses shell script to build whole project) to see changes in completion.
And it's very useful to see function parameters names in completion.

**Comment by ostinelli** [comment]
well, rebar is a requirement for sublimerl, so custom-made compile scripts are not supported (for testing functionalities, for instance). if source is used, this might complicate things as we need to reparse what rebar already does to find which .erl files have to be considered.

plus, getting function params might not be that easy: pattern matching complicates this:

```
myfun({test, Param1}, Param2) ->
    ok;
myfun(Param3, {test, Param4}) ->
    ok.
```

what params name would you get from here?

will dig into this idea too. thank you.

**Comment by tdx** [comment]
Yes, really in this case it is not easy, especially if there is binary matching. But if function has - spec then not so terribly.

**Comment by tdx** [comment]
For example 

<pre>
-spec accept(ListenSocket, Timeout) -> {ok, Socket} | {error, Reason} when
      ListenSocket :: socket(),
      Timeout :: timeout(),
      Socket :: socket(),
      Reason :: closed | timeout | system_limit | inet:posix().
</pre>


edoc transform this into

<pre>
accept(ListenSocket, Timeout) -> {ok, Socket} | {error, Reason}

    * ListenSocket = socket()
    * Timeout = timeout()
    * Socket = socket()
    * Reason = closed | timeout | system_limit | inet:posix()
</pre>


and completion may look something like 

accept(ListenSocket :: socket(), Timeout :: timeout()) 

or just

accept(ListenSocket, Timeout)

**Comment by ostinelli** [comment]
you can't count on spec presence though.

**Comment by ostinelli** [comment]
i've seen you closed the issue, you don't want this anymore?

i'm gonna think it through though, i think your suggestion needs to be taken into consideration and may lead to interesting results.

**Comment by tdx** [comment]
Closed because thought that isn't interesting to you.

It's would be excellent to have at least names of parameters in completion.

For functions which have no specification it is possible to show completion as now.

**Comment by tdx** [comment]
And one more plus for parse source code - it is possible to show help for selected function in completion.
For functions which have edoc tags.

**Comment by ostinelli** [comment]
i am giving this a shot, at first it will be simple but we might extend later on with functionalities. not sure if i like the idea that modules that cannot be compiled are part of code completion.

therefore: first test completion will not use edocs but variable names. will ping you when i have a working branch.

**Comment by tdx** [comment]
it seems current algorithm for regenerating completion is not optimal.
each time I'm swith to sublime editor the plugin starts regenerate project completion even if there are no changes - just switch off/on editor window.
on my machine regeneration takes about 2 minutes. this time cpu is about 80% load.

**Comment by ostinelli** [comment]
the generation of all erlang lib completions takes 45 seconds on my MBP, and this should happen only once per sublime text session.

so this is weird, since i doubt that you project is bigger than erlang code itself :)

are you Current\* files reasonable and contain only project info?

**Comment by tdx** [comment]
<pre>
  403317 2012-05-01 21:20 Current-Project.disasm
   30408 2012-05-01 21:20 Current-Project.sublime-completions
 2807719 2012-05-01 19:54 Erlang-Libs.disasm
  128091 2012-05-01 19:54 Erlang-Libs.sublime-completions
</pre>

**Comment by tdx** [comment]
Contents of files from the subdirectories rel/my_proj/lib/*/ebin are goes into disasm and completion.
Subdir rel is rebar generate command result and conatains part from erlang libs.
Whether there is a possibility to exclude some subdirectories from regeneration?

**Comment by ostinelli** [comment]
all of these are another issue. please continue the dicussion for this second part in the appropriate issues #19 or #20.

this issue should only treat about additional auto-completion information.

thank you,

r.

**Comment by ostinelli** [comment]
implemented: 2d8828f48055626e32ff25fa711bd55b7d8bc7f0

please let me know if everything is ok for you, and reopen the issue if not.

**Comment by tdx** [comment]
Some params not parsed.

Ex:

<pre>
-spec make_client_record(Usr::binary(), Opts::[atom()]) -> #client{}.

make_client_record(Usr, Opts) when is_binary(Usr),
                              is_list(Opts) ->
   ...
</pre>


completion:

<pre>
my_mod:make_client_record(Param1, Param2)
</pre>

**Comment by tdx** [comment]
can't reopen issue - You don't have permission to reopen this issue.

**Comment by tdx** [comment]
also not parsed binary and record params

fun1( << _:3/bytes,Body/binary >> = Data) 

completion:

<pre>
fun1(Param1)
</pre>


Records:

<pre>
fun1(#client{name=Name} = Client) ->
</pre>


Completion:

<pre>
fun1(Param1)
</pre>

**Comment by tdx** [comment]
module file has only one completion

file:format_error(Reason)

**Comment by ostinelli** [comment]
Thank you for this exhaustive report.

1.
`make_client_record(Usr, Opts) when is_binary(Usr), is_list(Opts) ->`

this is a bug, will fix.

2.
`fun1( << _:3/bytes,Body/binary >> = Data) ->`
`fun1(#client{name=Name} = Client) ->`

Currently the parser just understands plain name variables, so more complex things like these tuples, binary matching, records, etc are not intended to be parsed yet, and default do `Param1, Param2, ...`. However, these specific ones have a single variable assignment, so I will try to integrate these at least. 

Also, the parser is ignoring `-spec` directives right now. I am planning to integrate that, not sure when.

3.
`module file has only one completion`

will dig into this.

**Comment by ostinelli** [comment]
this should fix it, along with a considerable speed improvement: 833abdd35dc281e6519f9246607712ccdd87beca

please confirm, and if necessary reopen the issue :)

r.

**Comment by tdx** [comment]
Yes, it works fine for me.

**Comment by ostinelli** [comment]
Nice, thank you.
